### ISSUE TITLE
OSA [1563448] - Change to linuxbridge agent defaults affects upgrades

### ISSUE THREAD

**Issue by git-harry** [rcbops/rpc-openstack#973]
##### What version (commit) do you encounter this issue on?
master
##### What type of issue is this (bug, feature, documentation, regression)?
bug
##### What is the expected behavior?
See https://bugs.launchpad.net/openstack-ansible/+bug/1563448
##### What is actually happening?
See https://bugs.launchpad.net/openstack-ansible/+bug/1563448
##### How can we reproduce this issue?
See https://bugs.launchpad.net/openstack-ansible/+bug/1563448
##### How can we verify your issue is addressed?
See https://bugs.launchpad.net/openstack-ansible/+bug/1563448
##### What is your evaluation on the importance of this issue, and why?
This issue is impacting upgrades and has the potential to cause additional downtime.
###### Remove branches (lines) you do not wish to target.
- [x] master https://github.com/rcbops/rpc-openstack/pull/1375
- [x] mitaka-13.0 https://github.com/rcbops/rpc-openstack/pull/1377
- [x] liberty-12.2 https://github.com/rcbops/rpc-openstack/pull/1379

**Comment by git-harry** [comment]
I think the best solution for r12.1.0 is to override the variable on upgrades to set [vxlan]/l2_population = True. This minimises the blocking effect of this issue on the release

For a future release, r13.1 at the latest, we need to create an automated upgrade path for changing this variable to the recommended default of False.

**Comment by prometheanfire** [comment]
this is going to be done in doc, making a note here

**Comment by tomc603** [comment]
I'm going to write up what situations Support may need to change this in and what will need to be changed. I'll pass it by @busterswt before sending to Docs to make sure it looks sane.

**Comment by b3rnard0** [comment]
@tomc603 How's this coming along? This is priority for r12.2.0 docs. Adding @erikmwilson as he'll be adding this to r12.2.0 release notes.

**Comment by tomc603** [comment]
I've sent over my notes to @erikmwilson, and he has noted he'll look them over tomorrow. If he needs, I'll gladly dig around more or add info where needed.

**Comment by b3rnard0** [comment]
@tomc603 What's the status on the documentation for this issue?

**Comment by tomc603** [comment]
@b3rnard0 I've sent everything over to @erikmwilson, and it seems to have passed a sniff test by a couple people. I'm going to mark this closed.

**Comment by git-harry** [comment]
Previous discussions left open the possibility of addressing this issue programmatically in a future release. I'm reopening the issue to allow for a discussion regarding potential code changes in later versions.

I will also remove the milestone and assignee now the documentation task is complete.

**Comment by tomc603** [comment]
@git-harry Yes, sorry! I wasn't sure if we wanted a new ticket for future work or not.

**Comment by git-harry** [comment]
Lowering the priority because there is a workaround in place.

**Comment by cloader89** [comment]
I am taking this over from @ngraf

**Comment by cloader89** [comment]
I see this was closed and re-opened, can anyone let me know if this is still  an issue or not? I see it was handled in docs. Do we need any programmatic fix?

**Comment by tomc603** [comment]
This will be related to #1149. If we disable l3ha by default, we will need to enable l2pop by default. This is the opposite of what we intended for the Mitaka release, but if l3ha is buggy and not working, we aren't left with many other choices.

**Comment by BjoernT** [comment]
@tomc603 Whats the background for this comment. L3HA can be disabled without having to enable l2pop, you can still leave l2pop disabled while using VXLAN multicast

**Comment by alextricity25** [comment]
Bjoern is right. I don't think we need l2pop on so long as we have a VXLAN multicast group, regardless of whether or not we have L3HA. There was a problem with having l2pop on with l3ha, but we don't need to worry about that if VXLAN multicast is enabled. The VXLAN multicast group was added here:
https://review.openstack.org/#/c/252100/5

**Comment by tomc603** [comment]
@BjoernT Sorry, looking at too many things at the same time. The confusion came from this change https://review.openstack.org/#/c/252100/5/playbooks/roles/os_neutron/templates/neutron.conf.j2 which I just glanced at very quickly and didn't grok properly.

Anyway, our initial intent was to use a VXLAN multicast group to do the L2 work.

**Comment by b3rnard0** [comment]
@git-harry Still need clarification on the technical direction for this issue and @cloader89 will need assistance once the decision is made. cc: @mancdaz

**Comment by mancdaz** [comment]
@b3rnard0: @tomc603 is taking ownership of getting this issue resolved, so he can help @cloader89 work through the final decisions.

**Comment by tomc603** [comment]
Ok, third time's a charm. I've drafted several comments on this issue and commented none of them.

@BjoernT if we enable vxlan multicast for l2 population purposes, then we will need to recreate the existing VXLAN interfaces so that existing VM can communicate properly with new VM created after upgrading.

This is a downtime event for each vm, and will require an agent restart as well. So if the environment is active during upgrade, the customer will notice an outage. We can do one of several things here:

* Orchestrate destroying interfaces, then restarting the agent one-at-a-time. This means the agent will be restarted over and over again, and as each VXLAN interface is recreated the others will lost contact with it until they also have their interfaces recreated. Not ideal for east-west traffic between VMs, but might have some advantages for north-south traffic. I'd want @busterswt or @Apsu to comment on this, though, to double check what I'm saying.

* Create a manually executed playbook which searches for effected VXLAN interfaces in the manner I described in the migration section of my upgrade notes to the docs team [1], deletes those interfaces, and when all interfaces are deleted restarts the agent. Restarting the agent, as noted, will recreate the VXLAN interfaces with the proper options. Making this a manually executed playbook will allow support to coordinate with our customers on the best time to perform this maintenance, keeping in mind that any _new_ VM will not be able to communicate with _existing_ VM in an east-west manner.

* Override the variable to continue doing l2 population. This is the least preferable option to me personally, because it keeps us behind on what is considered the preferred configuration from the Neutron project's perspective. It also postpones the inevitable migration to VXLAN multicast flooding L2 population, which will need to happen when we finally enable things like l3ha. The upshot of this is that customers are not impacted if they were previously not using VXLAN multicast population.

Let's set some time aside to discuss so we can come to a final conclusion on this, and get patches committed. Perhaps a meeting with @Apsu, @BjoernT and myself?

[1] https://docs.google.com/document/d/119DvoJVDmnzmp3h3ngFM4EN1HqnCc-WJsxPD2gOANLw/edit?usp=sharing

**Comment by tomc603** [comment]
I just saw @rcbops/support-feedback on another issue. Maybe someone on that list can provide feedback to my last comment?

**Comment by BjoernT** [comment]
@tomc603  What is the ask here? We have already automated the conversion to VXLAN multicast as part of the Liberty upgrade. As I said most customers will stay on l2pop after upgrade since multicast is mostly not working or we have not tested all environments and stay at l2pop.
We will not do an upgrade where we run the upgrade playbooks and change to VXLAN multicast at the same time. As a side fact, we are actually migrating bigger customers back to VLAN since we have stability and performance issues. We offer so far only, optionally, the move to vxlan multicast but most upgrades stayed at l2pop

**Comment by tomc603** [comment]
@BjoernT @rcbops/support-feedback Ok. Is it support's opinion we should go back to `l2_population = True` given the above comment?

**Comment by BjoernT** [comment]
correct

**Comment by tomc603** [comment]
I have created two PRs for this.

Master: https://github.com/rcbops/rpc-openstack/commit/d2ff1b40cec8208ac6e116aff7405c67f931d4b4

Liberty: https://github.com/rcbops/rpc-openstack/commit/614ba58304af3d0f3c69c1c722745d7ee6765468

Please review the release note content as well to be sure it clearly spells out concerns with potentially changing this default in existing environments.

**Comment by git-harry** [comment]
@tomc603 you forgot to create the pull requests. I've created one for the master commit. Once that is merged we can sort out the backports.

**Comment by evrardjp** [comment]
If you don't mind, I'll give my opinion based on @BjoernT  comment.

"What is the ask here? We have already automated the conversion to VXLAN multicast as part of the Liberty upgrade."
    Good. So I expect all the customers have a way to be with Liberty with l2pop disabled, as recommended by the community. So the safe way forward.

"As I said most customers will stay on l2pop after upgrade since multicast is mostly not working or we have not tested all environments and stay at l2pop."

    What does mostly not working mean? Do you have a bug report? Were the issues you encountered fixed in the latest version of the branch you're running? Or fixed in a newer branch? Don't you trust the problem could be solved by being more up to date? Do you need help in any way?

    Also, I guess this sentence would be true if the last part is correct ("We have not tested all environments and stay at l2pop"), so let's focus on this too: What do you need to trust this? What do you want to test? Could we help in any way?

"We will not do an upgrade where we run the upgrade playbooks and change to VXLAN multicast at the same time."

    Well that makes sense to reduce the amount of possible issues by reducing the scope of changes.

"As a side fact, we are actually migrating bigger customers back to VLAN since we have stability and performance issues."
    I am not sure what was meant here.

"We offer so far only, optionally, the move to vxlan multicast but most upgrades stayed at l2pop"
    Well that's an error, IMO. At some point support's gonna end up with more disparity, and every single cloud different. If you're willing to have an army of openstack administrators to do the work it's fine, but probably more expensive for the company.

    We should have one stance on this topic, which is stop using l2pop before it's too late.


Finally, I don't want to start a war and I want to avoid blood spill, I just want to ensure you guys have the proper support of the dev teams available to you.
Isolating yourselves in your island isn't a great solution in the long run, even if your island looks comfortable.

Let's not change anything in 13.0 or 13.1, and let's stay with l2pop disabled, as recommended by neutron guys.

**Comment by tomc603** [comment]
All PRs have merged. Closing this issue.

**Comment by git-harry** [comment]
@tomc603 status-pending-sha-update is only used when the fix depends on something upstream - what does this issue depend on?

**Comment by tomc603** [comment]
I misunderstood the lane's definition. Closing this issue again.
